### PR TITLE
Beta standardize builders2

### PR DESCRIPTION
**Pull request by PhilPJL**
Add another version of DefinitionBuilder that takes a TBuilder parameter.  I'm thinking about scripting all builders and this would be required to support that.  Also allows simplification of Configure.
Some refactor.

### CODE REVIEW

**Review by ThyWoof — DISMISSED** [review]

**Comment by ChrisPJohn** [comment]
I haven't had a chance to do a thorough review of this and I'd like to.

This is more a comment on the implied direction in the pull comment. I don't know if scripting builders is particularly useful. With extension support for setting all fields and the new configure, it seems the functionality that a scripted builder would provide is covered. The builders previously handled a lot of initialization logic that was needed as well as had some thought put into methods that set fields together which would commonly be set at the same time (like these 3 fields are needed to properly set up a saving throw, so there's a SetSavingThrow method that has those parameters). Automatically created builders are not going to have that.

**Review by ChrisPJohn — CHANGES_REQUESTED** [review]

File: SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerBuilder.cs [hunk 1]
```diff
@@ -1,56 +1,63 @@
...
+            return new FeatureDefinitionPowerBuilder(name, guid);
+        }
 
         public static FeatureDefinitionPowerBuilder Create(string name, Guid namespaceGuid)
         {
             return new FeatureDefinitionPowerBuilder(name, namespaceGuid);
         }
 
+        // Over specific method?
+        public FeatureDefinitionPowerBuilder Configure(
```
[the remark this answers is not in the document]
Comment (PhilPJL) [reply SolastaCommunityExpansion/Builders/Features/FeatureDefinitionPowerBuilder.cs:41]: Ok understood.

File: SolastaCommunityExpansion/Classes/Tinkerer/Subclasses/BattleSmithBuilder.cs [hunk 2]
```diff
@@ -13,12 +13,9 @@ public static class BattleSmithBuilder
         public static CharacterSubclassDefinition Build(CharacterClassDefinition artificer)
         {
             // Make Battle Smith subclass
-            CharacterSubclassDefinitionBuilder battleSmith = new CharacterSubclassDefinitionBuilder("BattleSmith", GuidHelper.Create(TinkererClass.GuidNamespace, "BattleSmith").ToString());
-            GuiPresentationBuilder battleSmithPresentation = new GuiPresentationBuilder(
-                "Subclass/&ArtificerBattleSmithTitle",
-                "Subclass/&ArtificerBattleSmithDescription");
-            battleSmithPresentation.SetSpriteReference(DatabaseHelper.CharacterSubclassDefinitions.MartialSpellblade.GuiPresentation.SpriteReference);
```
[the remark this answers is not in the document]
Comment (PhilPJL) [reply SolastaCommunityExpansion/Classes/Tinkerer/Subclasses/BattleSmithBuilder.cs:20]: It looks like there never was one.

[the remark this answers is not in the document]
Comment (PhilPJL) [reply SolastaCommunityExpansion/Classes/Tinkerer/Subclasses/BattleSmithBuilder.cs:20]: Done

**Comment by PhilPJL** [comment]
> I haven't had a chance to do a thorough review of this and I'd like to.
> 
> This is more a comment on the implied direction in the pull comment. I don't know if scripting builders is particularly useful. With extension support for setting all fields and the new configure, it seems the functionality that a scripted builder would provide is covered. The builders previously handled a lot of initialization logic that was needed as well as had some thought put into methods that set fields together which would commonly be set at the same time (like these 3 fields are needed to properly set up a saving throw, so there's a SetSavingThrow method that has those parameters). Automatically created builders are not going to have that.

My intention - if it's thought worthwhile and I have enough energy - would be to create the complete hierarchy of builders (still haven't got it straight in my head) so may need two instances of each builder type like
    public abstract class FeatureDefinitionFeatureSetBuilder<TDefinition, TBuilder> : DefinitionBuilder<TDefinition, TBuilder>
and
    public class FeatureDefinitionFeatureSetBuilder : FeatureDefinitionFeatureSetBuilder<FeatureDefinitionFeatureSet, FeatureDefinitionFeatureSetBuilder>
I may be overthinking things.

And then have all of the definition setters as private or protected members.
These can be promoted to public as necessary - saves typing - or used in higher level methods.

**Comment by PhilPJL** [comment]
Apologies.  I thought I had branched to make further changes on top of previous PR.  When I did a push they added to this one.

**Review by ChrisPJohn — APPROVED** [review]

**Comment by PhilPJL** [comment]
I should have mentioned that if I did create a script this would be used once only not on every release. I'm not sure it's worth creating since we probably have builders for the most used definitions already.

**Pull request merged**